### ISSUE TITLE
C++ Style Guide: 64-bit Portability conflicts with Integer Types section

### ISSUE THREAD

**Issue by mattgruenke** [google/styleguide#179]
Under 64-bit Portability, the guide advises:

> Use the LL or ULL suffixes as needed to create 64-bit constants.

Yet, the Integer Types section states:

> stdint.h defines types like int16_t, uint32_t, int64_t, etc. You should always use those in preference to short, unsigned long long and the like, when you need a guarantee on the size of an integer.

Therefore, it seems that the 64-bit Portability section should instead advise developers to use the INTn_C() and UINTn_C() macros, defined in stdint.h (or cstdint).  See:

```
http://en.cppreference.com/w/c/types/integer
http://en.cppreference.com/w/cpp/header/cstdint
```

BTW, if any rationale was given to prefer stdint.h over cstdint, I seem to have missed it.

**Comment by TafThorne** [comment]
Does the part about using LL and ULL not specifically mean constants rather than integers?  

So it is talking about replacing `#define THING 123` or `const int THING = { 123 };` with `LL THING 123` rather than replacing `int thing = { 321 };` with `int64_t thing = { 321 };`.

**Comment by tituswinters** [comment]
Don't use the macros, those are terrible.

The integer types section is more current than the portability section.  I'd definitely suggest initialization of the form:

int64_t thing = 123;

or if that doesn't work in context,
int64_t thing{other};
(but only as a last resort).

**Comment by TafThorne** [comment]
My point was that constant is a reference to `const`.  I would agree that `const int64_t THING = 123;`  would be preferable in most C++ code but I do not control the Google codebase.  They value consistency between projects more than the ideal for today.  If they have very old code they may really feel it better to keep the macros.

**Comment by TafThorne** [comment]
Ops.  Could have checked the profiles and seen that tituswinters may be able to speak from a Google opinion perspective as a staff member.  At least my bit about const being relevant still stands.

**Comment by tituswinters** [comment]
Other than "don't use #define to declare a constant" I don't think constant-ness has much to do with it.  Globals and local and const and non-const, we'd say roughly the same thing.

Also, as an aside: the consistency argument is mostly local consistency these days. There's too much change when you look at the language over the course of decades - arguing for complete consistency stopped being feasible a few years ago.

**Comment by TafThorne** [comment]
TLDR: I see the point about constant not being critical.  Then I waffle. 

When I ready the [section in question](https://github.com/google/styleguide/blob/gh-pages/cppguide.html#L3117) I can see that the word constant is probably not refering to const at all even though the preding paragraphs are about macro defines (which would be constant).  It sounds like the whole section has fallen out of date with current practise at Google and current trends in C++ style.  More generally the rule of _never use #define_ (unless there really is no alternative) seems to stand.  Even if it means it looks odd compared to C.  If it were to get updated I would suggest the word constant be dropped or the keyword `const` be added.  

Glad to hear that Google are being pragmatic about their code bases.  Lots of old articles still exist about how great the consistency is for moving between teams.  I can see the point but it would be very hard to keep everything set at... what an early to mid 2000 style for C & C++.  Balance in all things.

**Comment by tituswinters** [comment]
There's a lot of consistency among current-development projects, it's just no longer practical to require consistency across the entire codebase including legacy stuff.  Practices advance - if we required complete consistency we'd be stuck in a world without std::unique_ptr.

I'll try to carve out some time to polish up the 64-bit section.  (I suspect it's largely a matter of deletion at this point, but it'll take a bit of thinking.)

Anyway, thanks for the discussion.

**Comment by mattgruenke** [comment]
I don't understand how this discussion veered into the long-settled debate of **const** vs. **#define**.  I thought it was clear that the issue was concerning how to indicate the type of a literal, which can sometimes affect the result of expressions in which they're used.  I've also seen compilers issue warnings when the value of a numeric literal lacking a type specifier exceeds the range representable by **int**.

As for macros, I think we can all agree that they should be avoided, except when nothing else can do the job.  This is one such case, which is clearly why the C++ standards committee included the **INTn_C()** and **UINTn_C()** macros in the standard library.

Just to be clear, the concern is that **LL** and **ULL** are non-portable, because they depend on the size of **long long**.  When one wants to specify a 64-bit signed or unsigned literal, it seems that **INT64_C( _number_ )** and **UINT64_C( _number_ )**, respectively, would both offer greater portability and better indicate the intent of the author.

**Comment by tituswinters** [comment]
I finally took a crack at updating the guidance for this to be in-line with more portable/more modern practice.  It'll be a little while before I get the public version of the guide posted (hopefully next week), but the summary/preview runs thusly:

On Unsigned Integers

Unsigned integers are good for representing bitfields and modular arithmetic. Because of historical accident, the C++ standard also uses unsigned integers to represent the size of containers - many members of the standards body believe this to be a mistake, but it is effectively impossible to fix at this point. The fact that unsigned arithmetic doesn't model the behavior of a simple integer, but is instead defined by the standard to model modular arithmetic (wrapping around on overflow/underflow), means that a significant class of bugs cannot be diagnosed by the compiler. In other cases, the defined behavior impedes optimization.

That said, mixing signedness of integer types is responsible for an equally large class of problems. The best advice we can provide: try to use iterators and containers rather than pointers and sizes, try not to mix signedness, and try to avoid unsigned types (except for representing bitfields or modular arithmetic). Do not use an unsigned type merely to assert that a variable is non-negative.

64-bit Portability
Code should be 64-bit and 32-bit friendly. Bear in mind problems of printing, comparisons, and structure alignment.

printf() specifiers for some types are not cleanly portable between 32-bit and 64-bit systems. Technically this can be resolved by using the PRI family of format string macros. However, we find these to be unpleasant to use and impractical to require. Unfortunately, the standard 64-bit types int64_t and uint64_t don't have good native format specifiers, only messy PRI format macros. Avoid passing values of any of these 64-bit types (or any other numeric typedef) to printf(). 

Use braced-initialization as needed to create 64-bit constants. For example:
int64 my_value{0x123456789};
uint64 my_mask{3ULL << 48};

**Comment by mattgruenke** [comment]
As I stated in my original comment, your use of the `ULL` suffix is inconsistent with guidance, in [Google C++ Style Guide: Integer Types](https://google.github.io/styleguide/cppguide.html#Integer_Types), to avoid variable types of `long long`.

> `<stdint.h>` defines types like `int16_t`, `uint32_t`, `int64_t`, etc. You should always use those in preference to `short`, `unsigned long long` and the like, when you need a guarantee on the size of an integer.

Why shouldn't the same heuristic apply to literals?  Why not use the standard `UINT64_C()` macro, instead?
